### PR TITLE
MAR-4-create-controller-with-CRUD-operations-for-Attribute

### PR DESCRIPTION
**Pull request by NikolayJRStudent**

### CODE REVIEW

File: src/test/java/ua/tc/marketplace/service/impl/AttributeServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    @Test
+    void deleteById_ShouldThrowException_WhenAttributeIsLinkedToCategory() {
+
+        when(attributeRepository.existsById(1L)).thenReturn(true);
+
+        when(categoryRepository.existsByAttributes_Id(1L)).thenReturn(true);
+
+        AttributeDeletionException exception = assertThrows(AttributeDeletionException.class, () -> attributeService.deleteById(1L));
+        assertEquals("Неможливо видалити атрибут, який зв'язаний з існуючою категорією", exception.getMessage());
```
Comment (OleksandrProkopenkoDev) [inline src/test/java/ua/tc/marketplace/service/impl/AttributeServiceImplTest.java:150]: translate into English, please)))

File: src/main/java/ua/tc/marketplace/service/impl/AttributeServiceImpl.java [hunk 2]
```diff
@@ -1,22 +1,99 @@
...
+    if (isAttributeLinkedToCategory(id)) {
+      throw new AttributeDeletionException();
+    }
+
+    attributeRepository.deleteById(id);
+  }
+
+  private boolean isAttributeLinkedToCategory(Long attributeId) {
+
+    return categoryRepository.existsByAttributes_Id(attributeId);
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/service/impl/AttributeServiceImpl.java:95]: can you make method name in pure camel case without `_`?

File: src/main/java/ua/tc/marketplace/controller/AttributeController.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+@RequestMapping("/api/v1/attribute")
+@RequiredArgsConstructor
+@Slf4j
+public class AttributeController {
+
+    private final AttributeService attributeService;
+
+    @GetMapping
+    public ResponseEntity<List<AttributeDto>> getAllAttributes(Pageable pageable) {
+        log.info("Request to get all attributes");
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/controller/AttributeController.java:32]: leave only one `log` per each  request.

File: src/main/java/ua/tc/marketplace/service/impl/AttributeServiceImpl.java [hunk 4]
```diff
@@ -1,22 +1,99 @@
...
+
+  @Transactional
+  @Override
+  public void deleteById(Long id) {
+    if (!attributeRepository.existsById(id)) {
+      throw new AttributeNotFoundException(id);
+    }
+
+    if (isAttributeLinkedToCategory(id)) {
+      throw new AttributeDeletionException();
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/service/impl/AttributeServiceImpl.java:87]: include attributeId in exception message

File: src/main/java/ua/tc/marketplace/controller/AttributeController.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    @GetMapping("/{id}")
+    public ResponseEntity<AttributeDto> getAttributeById(@PathVariable Long id) {
+        log.info("Request to get attribute with ID: {}", id);
+        AttributeDto attributeDto = attributeService.findById(id);
+        log.info("Attribute with ID: {} was retrieved successfully", id);
+        return ResponseEntity.ok(attributeDto);
+    }
+
+    @PostMapping
+    public ResponseEntity<AttributeDto> createAttribute(@Valid @RequestBody CreateAttributeDTO attributeDTO) {
```
Comment (vosmerkin) [inline src/main/java/ua/tc/marketplace/controller/AttributeController.java:47]: You're using different styles for abbreviations in class and variables names (AttributeDto and CreateAttributeDTO). Even though both are correct, you need to choose one. All previosly created Dtos has this style (AttributeDto).

File: src/main/java/ua/tc/marketplace/exception/attribute/AttributeDeletionException.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package ua.tc.marketplace.exception.attribute;
+
+import org.springframework.http.HttpStatus;
+import ua.tc.marketplace.exception.model.CustomRuntimeException;
+
+public class AttributeDeletionException extends CustomRuntimeException {
+
+    private static final String ERROR_MESSAGE = "Неможливо видалити атрибут, який зв'язаний з існуючою категорією";
```
Comment (vosmerkin) [inline src/main/java/ua/tc/marketplace/exception/attribute/AttributeDeletionException.java:8]: All other error messages are on English. Why did you changed to Ukrainian?

File: src/main/java/ua/tc/marketplace/controller/AttributeController.java (hunk 5 above)
Comment (OleksandrProkopenkoDev) [reply src/main/java/ua/tc/marketplace/controller/AttributeController.java]: actually I dont understand what do you mean))) I looked at the user dto package and it has the same name structure that are here in attributes))

But i found another questionable place: there are two exactly the same attribute dto with different names (CreateAttributeDto and UpdateAttributeDto). In some cases it is possible to make like this. But in our case i renamed one of them into `AttributeRequest` and deleted other one

**Review by OleksandrProkopenkoDev — APPROVED** [review]

**Pull request merged**